# Celebal Technologies Week 1 Assignment - Shaik Faizan Ahmed

## Question 1: Load a CSV dataset into a Pandas DataFrame.

In [1]:
import pandas as pd
df = pd.read_csv("Combined_dataset.csv")
print("Dataset loaded successfully")

Dataset loaded successfully


In [2]:
df

,product_id,title,product_description,rating,ratings_count,initial_price,discount,final_price,currency,images,...,amount_of_stars,what_customers_said,seller_name,sizes,videos,seller_information,variations,best_offer,more_offers,category
0,8376765,Lino Perros,Women Navy Blue Solid Backpack,3.8,15,3995,58.0,"""₹3,995.00""",INR,http://assets.myntassets.com/assets/images/837...,...,"{""1_star"":2,""2_stars"":0,""3_stars"":3,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""rw-8376765_cae700""]",NaN,[{}],{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
1,9136281,Tommy Hilfiger,Unisex Navy Blue Striped Backpack,4.5,67,2899,35.0,"""₹2,899.00""",INR,http://assets.myntassets.com/assets/images/913...,...,"{""1_star"":3,""2_stars"":4,""3_stars"":2,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""rw-9136281_cae700""]",NaN,"[{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
2,17633752,Lavie,Aries Women Pink Mini Backpack,4.4,226,2999,65.0,"""₹2,999.00""",INR,http://assets.myntassets.com/assets/images/176...,...,"{""1_star"":9,""2_stars"":5,""3_stars"":10,""4_stars""...",NaN,NaN,"[{""size"":""S""}]","[""https://videos.myntassets.com/assets/videos/...",NaN,"[{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
3,1376949,F Gear,Unisex Navy & Grey Printed Burner Backpack,4.4,1052,1675,52.0,"""₹1,675.00""",INR,http://assets.myntassets.com/assets/images/137...,...,"{""1_star"":33,""2_stars"":21,""3_stars"":69,""4_star...",NaN,NaN,"[{""size"":""Onesize""}]","[""https://videos.myntassets.com/assets/videos/...",NaN,NaN,{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
4,13939916,MYTRIDENT,Men Blue Solid Bath Robe,4.7,12,2899,17.0,"""₹2,899.00""",INR,http://assets.myntassets.com/assets/images/pro...,...,"{""1_star"":0,""2_stars"":0,""3_stars"":0,""4_stars"":...",NaN,NaN,"[{""size"":""M"",""value"":""45.0in"",""value_name"":""Fr...",NaN,NaN,"[{},{},{},{},{},{},{},{},{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on IndusI...",bath-robe
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,13698258,KREDO,Men FORTE Silver Automatic Skeleton Analogue W...,3.5,11,11890,50.0,"""₹11,890.00""",INR,http://assets.myntassets.com/assets/images/136...,...,"{""1_star"":3,""2_stars"":0,""3_stars"":2,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""pi-13698258""]",NaN,NaN,{},"[{""offer_name"":""10% Instant Discount on Citi C...",watches
996,16896006,GUESS,Women White Patterned Dial & White Straps Anal...,5.0,22,8900,45.0,"""₹8,900.00""",INR,http://assets.myntassets.com/assets/images/168...,...,"{""1_star"":0,""2_stars"":0,""3_stars"":0,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""rw-16896006_cae700"",""https://videos.myntasse...",NaN,NaN,{},"[{""offer_name"":""10% Instant Discount on Citi C...",watches
997,1708108,GIORDANO,Women Rose Gold-Toned Mirrored Analogue Watch ...,2.8,5,7950,50.0,"""₹7,950.00""",INR,http://assets.myntassets.com/assets/images/170...,...,"{""1_star"":2,""2_stars"":1,""3_stars"":0,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]",NaN,NaN,"[{},{},{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",watches
998,19280114,Fastrack,Men Black Brass Dial & Black Stainless Steel S...,4.3,17,3295,20.0,"""₹2,636.00""",INR,http://assets.myntassets.com/assets/images/192...,...,"{""1_star"":1,""2_stars"":2,""3_stars"":0,""4_stars"":...",NaN,Tru███m R███il,"[{""size"":""Onesize""}]",NaN,Sho███o. ███-10█████████enu███████████████████...,NaN,"{""applicable_on"":""Orders above Rs. 1499 (only ...","[{""offer_name"":""10% Instant Discount on IndusI...",watches


In [3]:
df.head()

,product_id,title,product_description,rating,ratings_count,initial_price,discount,final_price,currency,images,...,amount_of_stars,what_customers_said,seller_name,sizes,videos,seller_information,variations,best_offer,more_offers,category
0,8376765,Lino Perros,Women Navy Blue Solid Backpack,3.8,15,3995,58.0,"""₹3,995.00""",INR,http://assets.myntassets.com/assets/images/837...,...,"{""1_star"":2,""2_stars"":0,""3_stars"":3,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""rw-8376765_cae700""]",NaN,[{}],{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
1,9136281,Tommy Hilfiger,Unisex Navy Blue Striped Backpack,4.5,67,2899,35.0,"""₹2,899.00""",INR,http://assets.myntassets.com/assets/images/913...,...,"{""1_star"":3,""2_stars"":4,""3_stars"":2,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""rw-9136281_cae700""]",NaN,"[{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
2,17633752,Lavie,Aries Women Pink Mini Backpack,4.4,226,2999,65.0,"""₹2,999.00""",INR,http://assets.myntassets.com/assets/images/176...,...,"{""1_star"":9,""2_stars"":5,""3_stars"":10,""4_stars""...",NaN,NaN,"[{""size"":""S""}]","[""https://videos.myntassets.com/assets/videos/...",NaN,"[{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
3,1376949,F Gear,Unisex Navy & Grey Printed Burner Backpack,4.4,1052,1675,52.0,"""₹1,675.00""",INR,http://assets.myntassets.com/assets/images/137...,...,"{""1_star"":33,""2_stars"":21,""3_stars"":69,""4_star...",NaN,NaN,"[{""size"":""Onesize""}]","[""https://videos.myntassets.com/assets/videos/...",NaN,NaN,{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
4,13939916,MYTRIDENT,Men Blue Solid Bath Robe,4.7,12,2899,17.0,"""₹2,899.00""",INR,http://assets.myntassets.com/assets/images/pro...,...,"{""1_star"":0,""2_stars"":0,""3_stars"":0,""4_stars"":...",NaN,NaN,"[{""size"":""M"",""value"":""45.0in"",""value_name"":""Fr...",NaN,NaN,"[{},{},{},{},{},{},{},{},{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on IndusI...",bath-robe


### My Observation:
The dataset was successfully loaded into a Pandas DataFrame. The DataFrame will be used for data exploration, cleaning, and transformation in the upcoming steps.

## Question 2: Data Exploration (head/tail, shape, columns, data types)

In [4]:
df.head()

,product_id,title,product_description,rating,ratings_count,initial_price,discount,final_price,currency,images,...,amount_of_stars,what_customers_said,seller_name,sizes,videos,seller_information,variations,best_offer,more_offers,category
0,8376765,Lino Perros,Women Navy Blue Solid Backpack,3.8,15,3995,58.0,"""₹3,995.00""",INR,http://assets.myntassets.com/assets/images/837...,...,"{""1_star"":2,""2_stars"":0,""3_stars"":3,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""rw-8376765_cae700""]",NaN,[{}],{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
1,9136281,Tommy Hilfiger,Unisex Navy Blue Striped Backpack,4.5,67,2899,35.0,"""₹2,899.00""",INR,http://assets.myntassets.com/assets/images/913...,...,"{""1_star"":3,""2_stars"":4,""3_stars"":2,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""rw-9136281_cae700""]",NaN,"[{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
2,17633752,Lavie,Aries Women Pink Mini Backpack,4.4,226,2999,65.0,"""₹2,999.00""",INR,http://assets.myntassets.com/assets/images/176...,...,"{""1_star"":9,""2_stars"":5,""3_stars"":10,""4_stars""...",NaN,NaN,"[{""size"":""S""}]","[""https://videos.myntassets.com/assets/videos/...",NaN,"[{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
3,1376949,F Gear,Unisex Navy & Grey Printed Burner Backpack,4.4,1052,1675,52.0,"""₹1,675.00""",INR,http://assets.myntassets.com/assets/images/137...,...,"{""1_star"":33,""2_stars"":21,""3_stars"":69,""4_star...",NaN,NaN,"[{""size"":""Onesize""}]","[""https://videos.myntassets.com/assets/videos/...",NaN,NaN,{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
4,13939916,MYTRIDENT,Men Blue Solid Bath Robe,4.7,12,2899,17.0,"""₹2,899.00""",INR,http://assets.myntassets.com/assets/images/pro...,...,"{""1_star"":0,""2_stars"":0,""3_stars"":0,""4_stars"":...",NaN,NaN,"[{""size"":""M"",""value"":""45.0in"",""value_name"":""Fr...",NaN,NaN,"[{},{},{},{},{},{},{},{},{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on IndusI...",bath-robe


In [5]:
df.tail()

,product_id,title,product_description,rating,ratings_count,initial_price,discount,final_price,currency,images,...,amount_of_stars,what_customers_said,seller_name,sizes,videos,seller_information,variations,best_offer,more_offers,category
995,13698258,KREDO,Men FORTE Silver Automatic Skeleton Analogue W...,3.5,11,11890,50.0,"""₹11,890.00""",INR,http://assets.myntassets.com/assets/images/136...,...,"{""1_star"":3,""2_stars"":0,""3_stars"":2,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""pi-13698258""]",NaN,NaN,{},"[{""offer_name"":""10% Instant Discount on Citi C...",watches
996,16896006,GUESS,Women White Patterned Dial & White Straps Anal...,5.0,22,8900,45.0,"""₹8,900.00""",INR,http://assets.myntassets.com/assets/images/168...,...,"{""1_star"":0,""2_stars"":0,""3_stars"":0,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""rw-16896006_cae700"",""https://videos.myntasse...",NaN,NaN,{},"[{""offer_name"":""10% Instant Discount on Citi C...",watches
997,1708108,GIORDANO,Women Rose Gold-Toned Mirrored Analogue Watch ...,2.8,5,7950,50.0,"""₹7,950.00""",INR,http://assets.myntassets.com/assets/images/170...,...,"{""1_star"":2,""2_stars"":1,""3_stars"":0,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]",NaN,NaN,"[{},{},{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",watches
998,19280114,Fastrack,Men Black Brass Dial & Black Stainless Steel S...,4.3,17,3295,20.0,"""₹2,636.00""",INR,http://assets.myntassets.com/assets/images/192...,...,"{""1_star"":1,""2_stars"":2,""3_stars"":0,""4_stars"":...",NaN,Tru███m R███il,"[{""size"":""Onesize""}]",NaN,Sho███o. ███-10█████████enu███████████████████...,NaN,"{""applicable_on"":""Orders above Rs. 1499 (only ...","[{""offer_name"":""10% Instant Discount on IndusI...",watches
999,18601682,BOLDFIT,Purple Textured Yoga Mat,3.8,161,699,31.0,"""₹699.00""",INR,http://assets.myntassets.com/assets/images/186...,...,"{""1_star"":24,""2_stars"":8,""3_stars"":22,""4_stars...",NaN,NaN,"[{""size"":""Medium""}]",NaN,NaN,"[{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",yoga-mats


In [6]:
print("Rows and Columns:", df.shape)

Rows and Columns: (1000, 24)


In [7]:
print(df.columns)

Index(['product_id', 'title', 'product_description', 'rating', 'ratings_count',
       'initial_price', 'discount', 'final_price', 'currency', 'images',
       'delivery_options', 'product_details', 'breadcrumbs',
       'product_specifications', 'amount_of_stars', 'what_customers_said',
       'seller_name', 'sizes', 'videos', 'seller_information', 'variations',
       'best_offer', 'more_offers', 'category'],
      dtype='object')


In [8]:
print(df.dtypes)

product_id                  int64
title                      object
product_description        object
rating                    float64
ratings_count               int64
initial_price               int64
discount                  float64
final_price                object
currency                   object
images                     object
delivery_options           object
product_details            object
breadcrumbs                object
product_specifications     object
amount_of_stars            object
what_customers_said        object
seller_name                object
sizes                      object
videos                     object
seller_information         object
variations                 object
best_offer                 object
more_offers                object
category                   object
dtype: object


In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   product_id              1000 non-null   int64  
 1   title                   1000 non-null   object 
 2   product_description     1000 non-null   object 
 3   rating                  1000 non-null   float64
 4   ratings_count           1000 non-null   int64  
 5   initial_price           1000 non-null   int64  
 6   discount                879 non-null    float64
 7   final_price             1000 non-null   object 
 8   currency                1000 non-null   object 
 9   images                  1000 non-null   object 
 10  delivery_options        1000 non-null   object 
 11  product_details         1000 non-null   object 
 12  breadcrumbs             1000 non-null   object 
 13  product_specifications  1000 non-null   object 
 14  amount_of_stars         1000 non-null   o

#### My Observation:

- The dataset contains 1000 rows and 24 columns.
- Product information includes ratings, prices, sellers, offers, and categories.
- Some columns contain missing values.
- Most columns are stored as object (string) type.
- Further cleaning is required before analysis.

# Question 3: Handling Missing Values

In [10]:
df.isnull().sum()

product_id                  0
title                       0
product_description         0
rating                      0
ratings_count               0
initial_price               0
discount                  121
final_price                 0
currency                    0
images                      0
delivery_options            0
product_details             0
breadcrumbs                 0
product_specifications      0
amount_of_stars             0
what_customers_said       573
seller_name               301
sizes                       0
videos                    781
seller_information        301
variations                562
best_offer                  0
more_offers                 0
category                    0
dtype: int64

In [11]:
# Missing discount values are assumed to indicate no discount.
df["discount"] = df["discount"].fillna(0)

In [12]:
# Missing customer reviews indicate that no review information is available.
df["what_customers_said"] = df["what_customers_said"].fillna("No Reviews")

In [13]:
# Missing seller names are replaced with 'Unknown' to retain all records.
df["seller_name"] = df["seller_name"].fillna("Unknown")

In [14]:
# Missing video information indicates that no product video is available.
df["videos"] = df["videos"].fillna("No Video")

In [15]:
# Missing seller details are replaced with 'Not Available'.
df["seller_information"] = df["seller_information"].fillna("Not Available")

In [16]:
# Missing variation information indicates that no product variations are listed.
df["variations"] = df["variations"].fillna("No Variations")

In [17]:
df.isnull().sum()

product_id                0
title                     0
product_description       0
rating                    0
ratings_count             0
initial_price             0
discount                  0
final_price               0
currency                  0
images                    0
delivery_options          0
product_details           0
breadcrumbs               0
product_specifications    0
amount_of_stars           0
what_customers_said       0
seller_name               0
sizes                     0
videos                    0
seller_information        0
variations                0
best_offer                0
more_offers               0
category                  0
dtype: int64

### Missing Value Handling Summary

Missing values were identified in six columns. Instead of removing records, missing values were replaced with meaningful defaults based on the context of each attribute. This approach preserves all 1000 records while ensuring that no null values remain in the dataset.

# Question 4: Basic Operations

In [34]:
prod_summary = df[
    [
        "title",
        "category",
        "rating",
        "ratings_count",
        "initial_price",
        "discount"
    ]
]
prod_summary.head()

,title,category,rating,ratings_count,initial_price,discount
0,Lino Perros,backpacks,3.8,15,3995,58.0
1,Tommy Hilfiger,backpacks,4.5,67,2899,35.0
2,Lavie,backpacks,4.4,226,2999,65.0
3,F Gear,backpacks,4.4,1052,1675,52.0
4,MYTRIDENT,bath-robe,4.7,12,2899,17.0


In [35]:
high_rated_prod= df[df["rating"] >= 4.5]
high_rated_prod[
    ["title","rating","category"]
].head()

,title,rating,category
1,Tommy Hilfiger,4.5,backpacks
4,MYTRIDENT,4.7,bath-robe
5,H&M,4.5,bathroom-accessories
7,My Room,4.7,bedsheets
8,My Room,4.5,bedsheets


In [36]:
discounted_prod= df[df["discount"] > 0]
discounted_prod[
    ["title","discount","initial_price"]
].head()

,title,discount,initial_price
0,Lino Perros,58.0,3995
1,Tommy Hilfiger,35.0,2899
2,Lavie,65.0,2999
3,F Gear,52.0,1675
4,MYTRIDENT,17.0,2899


In [37]:
best_prod= df[
    (df["rating"] >= 4.5) &
    (df["discount"] > 0)
]
best_prod[
    ["title","rating","discount","category"]
].head()

,title,rating,discount,category
1,Tommy Hilfiger,4.5,35.0,backpacks
4,MYTRIDENT,4.7,17.0,bath-robe
7,My Room,4.7,54.0,bedsheets
8,My Room,4.5,54.0,bedsheets
15,Home Ecstasy,4.5,42.0,bedsheets


In [38]:
print("Highly Rated Products:", len(high_rated_prod))
print("Discounted Products:", len(discounted_prod))
print("Highly Rated & Discounted Products:", len(best_prod))

Highly Rated Products: 165
Discounted Products: 879
Highly Rated & Discounted Products: 143


### My Observations
Several filtering operations were performed to identify products of interest.

- Selected important attributes such as product title, category, rating, ratings count, price, and discount for focused analysis.
- Identified 165 highly rated products with ratings greater than or equal to 4.5.
- Found 879 products that were available with a discount.
- Identified 143 products that were both highly rated and discounted.
- These filtering operations help in identifying products that may be attractive to customers based on quality and pricing.

# Question 5: Remove Duplicates

In [23]:
print("Duplicate Rows:", df.duplicated().sum())

Duplicate Rows: 0


In [24]:
df = df.drop_duplicates()

In [25]:
print("Dataset Shape After Removing Duplicates:", df.shape)

Dataset Shape After Removing Duplicates: (1000, 24)


### My Observation

No duplicate records were found in the dataset. Therefore, the dataset size remained unchanged even after trying to remove duplicates.

# Question 6: Create a derived column (total_amount = price * quantity)

### Since the dataset does not contain explicit `price` and `quantity` columns, I derived them from the available data:

- `price` was extracted from the `final_price` column by removing currency symbols and converting it to a numeric format.
- `quantity` was derived from the number of available size variants listed in the `sizes` column.
- Finally, `total_amount` was calculated as the product of price and quantity.

In [26]:
df["price"] = (
    df["final_price"]
    .str.replace("₹", "", regex=False)
    .str.replace(",", "", regex=False)
    .str.replace('"', "", regex=False)
    .astype(float)
)

In [27]:
df[["final_price", "price"]].head()

,final_price,price
0,"""₹3,995.00""",3995.0
1,"""₹2,899.00""",2899.0
2,"""₹2,999.00""",2999.0
3,"""₹1,675.00""",1675.0
4,"""₹2,899.00""",2899.0


In [28]:
import ast
df["quantity"] = df["sizes"].apply(
    lambda x: len(ast.literal_eval(x))
)

In [29]:
df[["sizes", "quantity"]].head()

,sizes,quantity
0,"[{""size"":""Onesize""}]",1
1,"[{""size"":""Onesize""}]",1
2,"[{""size"":""S""}]",1
3,"[{""size"":""Onesize""}]",1
4,"[{""size"":""M"",""value"":""45.0in"",""value_name"":""Fr...",5


In [30]:
df["total_amount"] = df["price"] * df["quantity"]

In [31]:
df[["price", "quantity", "total_amount"]].head()

,price,quantity,total_amount
0,3995.0,1,3995.0
1,2899.0,1,2899.0
2,2999.0,1,2999.0
3,1675.0,1,1675.0
4,2899.0,5,14495.0


### My Observation

A numeric `price` column was created from the `final_price` field. A `quantity` column was derived from the number of available size variants. Using these columns, a new feature named `total_amount` was created to represent the total value of a product based on its price and available quantity options.

# Question 7: Save the Cleaned Dataset
After performing data cleaning, transformation, and feature engineering, the cleaned dataset is saved as a new CSV file for future analysis and processing.

In [32]:
# Save the cleaned dataset as a new CSV file.
df.to_csv("cleaned_dataset_Faizan.csv", index=False)
print("Cleaned dataset saved successfully.")

Cleaned dataset saved successfully.


In [33]:
# Display the shape of the final cleaned dataset.
print("Final Dataset Shape:", df.shape)

Final Dataset Shape: (1000, 27)


### My Observation

The cleaned dataset was successfully exported as a CSV file named `cleaned_dataset_Faizan.csv`. The final dataset contains the original product information along with newly derived features such as price, quantity, and total_amount, making it suitable for further analysis.